### ISSUE TITLE
Propose of changing/extending the API

### ISSUE THREAD

**Issue by marshall-lee** [dryrb/dry-constructor#1]
Hello!

I was already experimenting with my new gem https://github.com/marshall-lee/function_object when I realized there's `dry-constructor`. My function_object comes from another perspective: you basically want not only inject dependencies but also make a `call`. Someone would call my function object just a service object, someone would call it a lambda with the ability to use letrec (to use instance methods, I mean).

There's no README written yet, everything is in this little spec: https://github.com/marshall-lee/function_object/blob/master/spec/function_object_spec.rb

I think it'd be cool to delegate a construction step of `FunctionObject` to `Dry::Constructor`. However, current API does not allow to associate something with each argument. For example, `FunctionObject` allows to define a `default: -> { ... }` option for arguments. I also plan to implement a mixin that adds  `type: ` annotations to use the power of `dry-data` types. All of these things require another approach to describe arguments which are currently described just as a list of names.

I propose to bring the `FunctionObject`'s approach to describing a constructor:

```ruby
class CreateUserController
  include Dry::Constructor.define {
    arg :response
    arg :validator
    arg :command
  }
end
```

**Comment by AMHOL** [comment]
@marshall-lee there are plans to add support for kwargs if that's what you're suggesting? Not sure about the default at the moment, could be nice but will have to think about it. As for the type annotations you could delegate to something like https://github.com/egonSchiele/contracts.ruby for that?

**Comment by marshall-lee** [comment]
@AMHOL 

> As for the type annotations you could delegate to something like https://github.com/egonSchiele/contracts.ruby for that?

I know about contracts.ruby, but:

1) I want `dry-data`-driven annotations because `dry-data` rocks. It gives a well reasoned approach to type-checking problem in Ruby.
2) `contracts.ruby` is not suitable here. [Invalid object is an antipattern](http://solnic.eu/2015/12/28/invalid-object-is-an-anti-pattern.html), right? I want to prevent a construction of object with invalid dependencies so I want to do this check right in the constructor. The `initialize` method added by `Dry::Constructor` is hidden so I cannot even add type-annotations using `contracts.ruby` for it. `contracts.ruby` has a beautiful DSL but it does not fit here unfortunately.

TL;DR: I don't want a method type annotations, I want a *constructor* type annotations.

**Comment by marshall-lee** [comment]
@AMHOL 
> there are plans to add support for kwargs if that's what you're suggesting? Not sure about the default at the moment, could be nice but will have to think about it

In short, I suggest a way of describing arguments that could be extended. For example, one could implement `Dry::Constructor::Typed` that will add a `type: ` option:

```ruby
class Something
  include Dry::Constructor.define {
    arg :a, type: Types::Strict::String
    arg :b, type: Types::Integer
  }
end
```

**Comment by marshall-lee** [comment]
The defaults:

```ruby
class Something
  include Dry::Constructor.define {
    arg :a
    arg :b, default: -> { a + 1 }
  }
end
```

**Comment by marshall-lee** [comment]
I also love multiline definitions of arguments because git diffs become cleaner :) It's always very clear what argument is removed/added because each argument sits in its own line.

**Comment by jeremyf** [comment]
Extending @marshall-lee's proposed API, would the following make sense?

Is it something to expose each and all of the parameter options?

```ruby
class Something
  include Dry::Constructor.define {
    arg :a
    kwarg :b
    splat :args
    kwsplat :kargs
    block :configuration_proc
  }
end
```

Is there an assumption that you could include multiple `include Dry::Constructor.define` for a given class? If so the `splat`, `ksplat` and `block` declaration might wreak havoc.

**Comment by pabloh** [comment]
@jeremyf , what about a way to prevent nil values at the args?

**Comment by jeremyf** [comment]
@pabloh Perhaps splice into `dry-types`?.

```ruby
class Something
  include Dry::Constructor.define {
    arg :a, format: :str
    kwarg :b
    splat :args
    kwsplat :kargs
    block :configuration_proc
  }
end
```

**Comment by solnic** [comment]
Would be nice to figure out what to do with this library when there's also dry-initializer now. I'm afraid both libs will be confusing and people won't immediately understand which one should they use. In fact, I have that problem already. What should dry-constructor do? What should dry-initializer do? Should dry-initializer use dry-constructor and add type-related features on top?

**Comment by AMHOL** [comment]
FWIW I'm not keen on pushing the responsibility of coercion back to the object that the properties are on, it feels like moving back towards virtus. I've added the behaviour that I think dry-constructor should handle in [this PR](https://github.com/dry-rb/dry-constructor/pull/5) please check it out if you get a chance

**Comment by pabloh** [comment]
@jeremyf, I think `dry-initializer` is the most "feature complete" one, although I don't really see any other dry-rb gem depending on it.
